coreax/util.py

```diff
@@ -82,6 +83,30 @@ def apply_negative_precision_threshold(
     return x
 
 
+def pairwise(
+    fn: Callable[[ArrayLike, ArrayLike], Array],
+) -> Callable[[ArrayLike, ArrayLike], Array]:
+    """
+    Transform a function so it returns all pairwise evaluations of its inputs.
+
+    :param fn: the function to apply the pairwise transform to.
+    :returns: function that returns an array whose entries are the evaluations of `fn`
+        for every pairwise combination of its input arguments.
+    """
+
+    @ft.wraps(fn)
+    def pairwise_fn(x: ArrayLike, y: ArrayLike) -> Array:
+        x = jnp.atleast_2d(x)
+        y = jnp.atleast_2d(y)
+        return vmap(
```

Comment:
I'll investigate the performance regression and update when I have a good answer.

Reply:
Regression is due to removal of `jit` decoration of `Kernel.compute`. It is thought that the performance difference is due to enhanced compile times when compute is explicitly JIT'ed. It may be necessary to revisit this in the future, but as the degradation is not fundamental to this PR, I have reverted the removal of the compute `jit` decorator in the latest commit.